lightning/src/util/ser.rs

```diff
@@ -195,7 +195,7 @@ impl<R: Read> Read for ReadTrackingReader<R> {
 /// This is not exported to bindings users as we only export serialization to/from byte arrays instead
 pub trait Writeable {
 	/// Writes `self` out to the given [`Writer`].
-	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error>;
+	fn write(&self, writer: &mut impl Writer) -> Result<(), io::Error>;
```

Comment:
"except that the type is anonymous and doesn't appear in the [GenericParams](https://doc.rust-lang.org/reference/items/generics.html) list." 😁

Reply:
I don't understand what it does under the hood, but there is this note:
```
Note: For function parameters, generic type parameters and impl Trait are not exactly equivalent. With a generic parameter such as <T: Trait>, the caller has the option to explicitly specify the generic argument for T at the call site using [GenericArgs](https://doc.rust-lang.org/reference/paths.html#paths-in-expressions), for example, foo::<usize>(1). If impl Trait is the type of any function parameter, then the caller can't ever provide any generic arguments when calling that function. This includes generic arguments for the return type or any const generics.

Therefore, changing the function signature from either one to the other can constitute a breaking change for the callers of a function.
```

Reply:
Argh yeah, you are right, this approach does not work. 
I fixed all the places where the trait is being implemented and finally wrote a small test and it fails with:
```
--> lightning/src/util/ser.rs:198:5
     |
198  |     fn write(&self, writer: &mut impl Writer) -> Result<(), io::Error>;
     |        ^^^^^ ...because method `write` has generic type parameters
     |
    ::: lightning/src/sign/ecdsa.rs:170:11
```
In addition, I tested `impl Trait` in the return position and it still breaks object safety. No Christmas magic 😞